File: src/dhl/poll.go
package dhl

import (
	"lightspeed-dhl/config"
	"lightspeed-dhl/database"
	"lightspeed-dhl/lightspeed"
	"time"

	"github.com/rs/zerolog/log"
)

func StartPolling(conf *config.Secrets) {
	sleepDuration := time.Duration(conf.Options.PollingInterval) * time.Minute

	for {
		orders, err := database.GetUnprocessed()
		if err != nil {
			log.Err(err).Stack().Msg("Failed to get all orders")
			time.Sleep(sleepDuration)
			continue
		}

		log.Info().Int("entries_in_database", len(orders)).Msg("Polling for labels")
		for i := range orders {
			order := orders[i]

			// Create base of logging context
			baseLogger := log.With().Stack().Int("order_number", *order.LightspeedOrderId).Str("DHL draft id", *order.DhlDraftId)
			logger := baseLogger.Logger()

			logger.Debug().Interface("order", order).Msg("Processing order")

			// Check with DHL api if a label has been created for this order
			label, err := GetLabelByReference(*order.LightspeedOrderId, conf)
			if err != nil {
				logger.Err(err).Msg("Error getting label by reference")
				continue
			}

			if label == nil {
				logger.Debug().Msg("No label found")
				continue
			}

			// Add extra information to logging context
			logger = baseLogger.Str("order_reference", label.OrderReference).Logger()

			logger.Debug().Interface("label", label).Msg("Label found")
			// Set shipment id for this order in the database
			err = database.SetShipmentId(*order.DhlDraftId, label.ShipmentId)
			if err != nil {
				logger.Err(err).Msg("Error setting shipment id")
				continue
			}

			data, err := lightspeed.GetOrder(*order.LightspeedOrderId, conf)
			if err != nil {
				logger.Err(err).Msg("Error getting order from lightspeed, it might have been deleted")
				continue
			}

			isCancelled := data.Order.Status == "cancelled"
			logger.Debug().Bool("status", isCancelled).Msg("Order cancelled status")

			if isCancelled {
				logger.Info().Msg("Order is cancelled, removing from database")
				// Delete cancelled orders to prevent unnecessary iterations
				err := database.DeleteDraft(*order.DhlDraftId)
				if err != nil {
					logger.Err(err).Msg("Error deleting cancelled order from database")
				}
				continue
			}

			if !conf.Options.DryRun {
				err := lightspeed.UpdateOrderStatus(*order.LightspeedOrderId, lightspeed.UpdateOrderData{
					Status:         "completed_shipped",
					ShipmentStatus: "shipped",
				}, conf)
				if err != nil {
					logger.Err(err).Msg("Error updating order status")
					continue
				}

				logger.Debug().Msg("Order status updated to shipped")
			}
			// Set isProcessed in database
			err = database.SetProcessed(*order.DhlDraftId)
			if err != nil {
				logger.Err(err).Msg("Error setting processed")
				continue
			}
			logger.Debug().Msg("Order processed")
		}

		time.Sleep(sleepDuration)
	}
}

File: crates/adapter/src/poll.rs
use anyhow::{Context, Result};
use futures::StreamExt;
use tracing::{debug, error, info, warn};

use crate::{
    AdapterState,
    database::{self, models::Order},
};
use lightspeed::{OrderStatus, ShipmentStatus};

async fn reconcile_order_status(state: &AdapterState, order: &Order) -> Result<OrderStatus> {
    let order_id = order.lightspeed_order_id as u64;
    let order_number = order.lightspeed_order_number.clone();
    let draft_id = order
        .dhl_draft_id
        .context("order has been created without a draft id")?;

    database::incr_poll_count(&state.pool, order.lightspeed_order_id).await?;

    // we check for `order_id` first as this should be set as the order_reference.
    // labels may be created manually in which case this reference isn't set
    // by checking for the order_number there's a good chance we're still able to find it
    let label = match state.dhl.get_label(order_id).await {
        Ok(label) => Ok(label),
        Err(_) => state.dhl.get_label(&order_number).await,
    }?;

    if let Some(label) = label {
        info!({ shipment_id = ?&label.shipment_id }, "found label for order");

        let data = state
            .lightspeed
            .get_order(order_id)
            .await
            .context("unable to get lightspeed order for draft")?;

        if data.order.status.is_cancelled() {
            info!(
                { lightspeed_id = data.order.id },
                "order has been cancelled"
            );
            database::set_cancelled(&state.pool, &draft_id).await?;
            return Ok(OrderStatus::Cancelled);
        }

        database::set_shipment_id(&state.pool, &draft_id, label.shipment_id).await?;

        state
            .lightspeed
            .update_order_status(
                order_id,
                OrderStatus::CompletedShipped,
                ShipmentStatus::Shipped,
            )
            .await
            .context("unable to update lightspeed order status")?;

        let updated = state.lightspeed.get_order(order_id).await?.order;
        if !updated.is_shipped() {
            warn!(
                { status = ?&updated.status, shipment_status = ?&updated.shipment_status },
                "failed to update lightspeed order status to shipped"
            );
            return Ok(updated.status);
        }

        database::set_processed(&state.pool, &draft_id).await?;

        info!(
            { lightspeed_id = data.order.id, draft_id = ?&draft_id, shipment_id = ?&label.shipment_id },
            "order successfully processed"
        );

        return Ok(OrderStatus::CompletedShipped);
    }

    Ok(OrderStatus::ProcessingAwaitingPayment)
}

pub async fn run_once(state: AdapterState) -> Result<()> {
    let amount = database::unprocessed_count(&state.pool).await?;
    if amount == 0 {
        info!("nothing to do.");
        return Ok(());
    }

    info!("polling {amount} unprocessed orders");

    let mut orders = database::get_unprocessed_stream(&state.pool).await?;

    let mut processed: usize = 0;

    while let Some(query) = orders.next().await {
        match query {
            Ok(order) => match reconcile_order_status(&state, &order).await {
                Ok(status) => {
                    debug!(
                        { lightspeed_id = &order.lightspeed_order_id, dhl_id = ?&order.dhl_draft_id, status = ?&status },
                        "done checking order status"
                    );
                    if status == OrderStatus::CompletedShipped {
                        processed += 1;
                    }
                }
                Err(err) => error!(
                    { err = ?&err, order_id = &order.lightspeed_order_id  },
                    "failed to check order status"
                ),
            },
            Err(err) => {
                error!(
                    { error = &err as &dyn std::error::Error },
                    "unable to get next order from database"
                );
                continue;
            }
        };
    }

    if processed > 0 {
        info!({ processed_count = processed }, "done processing orders");
    }

    Ok(())
}
